docs/field_presence.md

```diff
@@ -0,0 +1,459 @@
+# Application note: Field presence
+
+This application note explains the various presence tracking disciplies for protobuf fields. It also explains how to enable experimental support for explicit presence tracking for singular proto3 fields with basic types.
+
+## Background
+
+_Field presence_ is the notion of whether a protobuf field has a value. Different types of fields follow different _presence disciplines_, or semantics for translating between the _API representation_ and the _serialized representation_.
+
+The _implicit_ presence discipline relies upon the field value itself to express notional presence. This can also be called _in-band_ presence tracking, since presence (or, usually, absence) is represented by an otherwise valid value.
```

Comment:
Can we simply say that such fields do not have presence? That is the terminology we have used previously:

- The original proto3 design (go/proto3-design) leads with:

> The main aspects that characterize Proto3 are the following: (1) Removal of field presence logic for primitive value fields [...]

- In several places throughout our code-base [we have methods](https://github.com/protocolbuffers/protobuf/search?q=HasFieldPresence&unscoped_q=HasFieldPresence) called `HasFieldPresence()`. This reinforces the idea that presence is something that fields have or don't.
- Based on these precedents I recently created FieldDescriptorProto::has_presence(), which indicates whether the field tracks presence.

If we change the terminology so that we now say that such fields have "implicit presence", I'm worried that the inconsistency will be confusing.

Reply:
Done... I updated everything to say either "no presence" or "explicit presence."